Makefile

```
@@ -40,7 +40,7 @@ build-preview: module-check ## Build site with drafts and future posts enabled
 	hugo --cleanDestinationDir --buildDrafts --buildFuture --environment preview
 
 deploy-preview: ## Deploy preview site via netlify
-	hugo --cleanDestinationDir --enableGitInfo --buildFuture --environment preview -b $(DEPLOY_PRIME_URL)
+	GOMAXPROCS=1 hugo --cleanDestinationDir --enableGitInfo --buildFuture --environment preview -b $(DEPLOY_PRIME_URL)
```

Should we apply this limit to _main_ branch builds as well?

---

```suggestion
	GOMAXPROCS=2 hugo --cleanDestinationDir --enableGitInfo --buildFuture --environment preview -b $(DEPLOY_PRIME_URL)
```
could this work OK too?

---

Not sure. The container we got from Netlify is 1C1G. I don't think relaxing this to 2 processors will buy us anything. I'm more inclined to start with 1 processor so that any concurrency bugs can be ruled out. Later on, if this proved to be working as expected, we can try relax the constraint.

---

Ah, OK. I am less worried about the previews and more worried about the main branch builds (which produce the live site).

Anyway, if this change is helpful, I support it. Do we know why it helps?

---

It is based on an observation from https://github.com/kubernetes/website/pull/36741. The site build (preview) seldom time out when I have 'GOMAXPROCS=1' specified. Actually, I wasn't able to reproduce the time out for the past 40 runs, no matter how I rebase or incorporate additional changes into that PR. When I removed 'GOMAXPROCS=1' today, I can reproduce the time out pretty soon.

Limiting CPU usage to one CPU will slow down the build a little bit. A normal build takes about 4 minutes. That is the tax we have to pay before we have a better idea how the frequent time out is caused.

---

https://github.com/golang/go/issues/33803 again 🙃 

we should consider PR-ing hugo to have https://github.com/uber-go/automaxprocs in the meantime

---

With throttling to 1 CPU, should setting GOMAXPROCS to 32 explicitly lead to build times going from 4 to over 60 minutes? We get timeouts when Hugo uses the automatic setting which we assume is 32.

I can picture how that might happen, but those scenarios feel more like bugs in something (maybe an old kernel?)

---

If netlify is hosting on a 32 core host but running in a container throttled to 1 core then Go will attempt to use 32 cores and will quickly exhaust the process's quota leading to aggressive throttling in the kernel. It's hard to say exactly what slowdown will result but it can be large.

If you control the execution environment it's often recommended to only set CPU requests (CFS shares -- proportional time ~guaranteed to schedule) and not limits (CFS quota -- time bound above which process is throttled) unless you're benchmarking. Linux cfs quota throttling is known to be heavy handed.

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/resource_management_guide/sec-cpu

I took a quick peek through hugo yesterday and AFAICT it's using go's default runtime.GOMAXPROCS() which would match the number of cores on the underlying host (roughly speaking) rather than the cfs quotas (see linked go issue as wel).

I'm not sure where best to insert it into hugo, I'd usually say in `main()` but they have a very tiny main, I'd guess this affects many other projects and hugo would generally benefit from defaulting GOMAXPROCS in a CFS aware way until go standard runtime solves this (as java did already).

---

filing an upstream issue / PR